src/computability/regular_expression_to_NFA/star.lean

```diff
@@ -0,0 +1,271 @@
+/-
+Copyright (c) 2022 Russell Emerine. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Russell Emerine
+-/
+import computability.regular_expression_to_NFA.defs
+
+/-!
+# Proof That Converting `r*` to NFA is Correct
+
+Proves that given that `r` converts to an NFA correctly, then `r*` converts to an NFA correctly.
+This is done by induction using the stronger condition that the number of "resets" in the machine
+matches the exponent in `r ^ n`.
+
+TODO: possibly merge the files in regular_expression_to_NFA together?
+-/
+
+universe u
+
+variables {α : Type u}
+
+namespace regular_expression
+
+/--
+Raises a regular expression to a power.
+
+TODO:
+ * probably move this into computability/regular_expression
+ * maybe change the order to `pow n * r`?
+-/
+def pow (r : regular_expression α) : ℕ → regular_expression α
+| 0 := epsilon
+| (nat.succ n) := r * pow n
+
+instance has_pow : has_pow (regular_expression α) ℕ := {pow := pow}
+
+theorem pow_def {r : regular_expression α} {n} : r.pow n = r ^ n := rfl
+
+theorem pow_succ {r : regular_expression α} {n : ℕ} {x} :
+  x ∈ (r ^ n.succ).matches ↔ x ∈ (r ^ n * r).matches :=
+begin
+  induction n generalizing x,
+  case zero
+  { split,
+    { rintros ⟨a, b, ha, hb, eq⟩,
+      cases hb,
+      refine ⟨list.nil, a, rfl, ha, _⟩,
+      simp [← eq], },
+    { rintros ⟨a, b, ha, hb, eq⟩,
+      cases ha,
+      refine ⟨b, list.nil, hb, rfl, _⟩,
+      simp [← eq], }, },
+  case succ : n ih
+  { split,
+    { rintros ⟨a, b, ha, hb, eq⟩,
+      rw [pow_def, ih] at hb,
+      rcases hb with ⟨c, d, hc, hd, eq'⟩,
+      refine ⟨a ++ c, d, language.append_mem_mul ha hc, hd, _⟩,
+      simp [eq, eq'], },
+    { rintros ⟨a, b, ha, hb, eq⟩,
+      rcases ha with ⟨c, d, hc, hd, eq'⟩,
+      refine ⟨c, d ++ b, hc, _, _⟩,
+      { rw [pow_def, ih],
+        exact language.append_mem_mul hd hb, },
+      { rw ← eq' at eq,
+        rw ← list.append_assoc,
+        exact eq, }, }, },
+end
+
+theorem star_iff_pow {r : regular_expression α} {x} :
+  x ∈ r.star.matches ↔ (∃ (n : ℕ), x ∈ (r ^ n).matches) :=
```

Comment:
I think this follows trivially from `language.star_eq_supr_pow`

Reply:
Thank you for all your feedback, Eric. I will be making changes and keeping an eye on #14261 .

Reply:
It's now merged :)